### ISSUE TITLE
Review Expired Listing Functionality

### ISSUE THREAD

**Issue by spencerfinnell** [Automattic/WP-Job-Manager#869]
Related #805, #822, #843, #860, #858, #865 

Tracking ticket to contain things about job expirations. 

There a couple of expectations on how expirations should work -- which will require adding additional options in the settings:

- A listing expires and it is no longer visible or accessible on the site at all.
  - This should set a status code of 404 like normal.
- A listing expires and still appears in listing results. Clicking the listing displays an "Expiration" notice.
  - This should set a status code of 401
- A listing expires and does not appear in listing results but the listing content itself is fully accessible (without the ability to apply).
  - This should set a status code of 200

Beyond that we need to ensure:

1. Dates input remain the same when saving/updating.
2. When the date occurs the item is properly expired baed on the chosen style above.

**Comment by Draxeiro** [comment]
Please note: the issue described in #858 is the same as issue #697. This bug was previously solved but since the update to Wordpress 4.7.1 it is back again.

Picking the date via the date picker will result in the listing date not being added properly (the date will show correctly until the listing is saved). The summary page will indeed show 1-1-1970 as the date and on the job listing page itself the date will be in the past (so the listing will not show any more).

Filling in the date by hand will work as a temporary workaround. Please be sure to fix this annoying bug which occurs for users on a non-English language version

**Comment by annezazuu** [comment]
affecting 3066029-t too

**Comment by rachelsquirrel** [comment]
Also reported in #3099049-t

I can reproduce this. I've got a job which has expired but still appears on my jobs page. I'm using the WC Paid Listings plugin as well.

![job](https://cloud.githubusercontent.com/assets/7946521/23895651/68b2d5d4-089e-11e7-9d0f-917c312a8290.png)

![job2](https://cloud.githubusercontent.com/assets/7946521/23895654/6bd5a264-089e-11e7-8cf4-3d3ceb39730e.png)

**Comment by bikedorkjon** [comment]
#3111459-t

**Comment by rachelsquirrel** [comment]
Also reported that when a payment fails the listing remains public. Reported in: #3102235-t

**Comment by spencerfinnell** [comment]
@rachelsquirrel Referring to a subscription payment failing?

**Comment by rachelsquirrel** [comment]
@spencerfinnell I believe so but not 100% sure. I'm checking.

**Comment by aaronjharms** [comment]
@spencerfinnell   Has there been any more progress on this Expiry Date issue?  Expiry dates in job editor are still moving back one day when saving.  But displaying correct expiry date in "All Jobs" list.

**Comment by spencerfinnell** [comment]
@jom this good to close?

**Comment by spencerfinnell** [comment]
@Suburblicious @aaronjharms This has been merged in to the master branch of the repository. You can download the latest version from Github for these fixes.

**Comment by jom** [comment]
@spencerfinnell #903 fixed a few issues (including the date input issue), but I'm not sure about the issue of listings not switching to expired after the expiration date. I'll try to dig into this today.

**Comment by jom** [comment]
I was able to reproduce a post being set to _expired_ after the expiration date. I'm closing this for now, but if it pops up again let's create a clean issue.

**Comment by spencerfinnell** [comment]
That actually through me off when I was testing @jom. I thin it's confusing. Opened https://github.com/Automattic/WP-Job-Manager/issues/974

**Comment by Suburblicious** [comment]
Greetings, should that be a fixed issue in the latest WP-job-manager update? I just updated to 1.26.0 and the problem is still present (but a bit different). I'm describing it here. Let me know if I should post somewhere else to make sure this gets resolved.

What happens:
- The employer sets an expiration date (i.e. June 18th) - with our settings, this date should be used as the "application closing date" and "listing expiry date" 
- When the admin views the job in the job listings, he sees June 18th as "closing date" and "expiry date"
- When the admin clicks on the job, the closing date is now minus one day (June 17th) and the application closing date is +3 days (June 21)
- When the admin saves the job WITHOUT CHANGING ANYTHING, the date on the website is now June 17th (minus one day from the employer's posting)

That is quite a major bug...

**Comment by jom** [comment]
@Suburblicious What theme are you using? Does it allow employers to choose the expiration date or are you referring to the automatic date set once a listing is approved?

Also, what time zone you are in?

**Comment by Suburblicious** [comment]
Greetings,

I'm using Workscout theme, with extension "WP Job Manager - Application Deadline", which lets the employer set an expiration date. Time zone is set to Toronto (-4 from UTC).

Normally using the website in French, but tested and reproduced the bug in English.

Thank you!

**Comment by jom** [comment]
Thanks, @Suburblicious. The application deadline changes the behavior of the job listing expiration by introducing a new meta field (`_application_deadline`) that (depending on your settings) can also expire a job listing. I've created a high priority issue to track this in the Application Deadline plugin.

**Comment by Suburblicious** [comment]
Thank you so much!

**Comment by Suburblicious** [comment]
Just to make sure: I deactivated the "Application deadline" extension. The problem with the "listing expiry date" is still present on the admin level: if the admin sets a date manually (June 18th), the dates switches to 17th (minus 1 day) upon saving the job posting.

To reproduce the bug:
- Create a job as an employer
- On the admin level, edit the job and choose a date for the Listing Expiry Date
- Save the job listing. Check the Listing Expiry Date: it is set for the day before the one you chose.

So the issue may be BOTH with the Application Deadline plugin AND the way WP-Job-Manager handles expiry dates...

**Comment by spencerfinnell** [comment]
Are you using version 1.26.0 from WordPres.org?

**Comment by Suburblicious** [comment]
Yes. Updated yesterday to WP-Job-manager 1.26.0 and expected the expiring issue to be fixed. 

Application deadline is version 1.1.5.

**Comment by jom** [comment]
Is the date you see in the Job Listing table in WP Admin different then the date you see when editing a job listing in WP Admin? 

I wonder if this may be caused by `var date = new Date( $(this).val() );` in `admin.js`. MDN has a notice that this will default to use UTC in some browsers.

**Comment by spencerfinnell** [comment]
Yeah something is still off. The placeholder is different than the preselected date: https://cloudup.com/cPp95oX7_pn

**Comment by Suburblicious** [comment]
Yes, if I visit the job listing table, the expiring date is 18 and if I click on the job to edit, it gets changed to 17th.

**Comment by spencerfinnell** [comment]
@jom Definitely related to the JS datepicker. The field value itself is correct until the JS takes over.

**Comment by jom** [comment]
I've opened a new issue #994 and referenced this conversation. Thanks for your help with this @Suburblicious.

(@spencerfinnell Looks like we had the same idea ;))

**Comment by Suburblicious** [comment]
OK, and would that explain the behavior for the Application deadline (which gets changed to +3 days)?

**Comment by jom** [comment]
@Suburblicious I'll have to do a bit more investigating. The Application Deadline plugin may override the WPJM core behavior and stores application due dates separately/in its own way.
